### PR TITLE
Properly handle init containers in convertToAPIContainerStatuses

### LINKED ISSUE #25844: Check container exit status when inspecting immediately after create
Kubelet issues a docker inspect call after creating a container, but for very short lived containers there is a window in which they could exit before being inspected [1]. Specifically, I'm referring to [this line](https://github.com/kubernetes/kubernetes/blob/35c9ca853a084a4ea04debe027c6ea680fd9ddac/pkg/kubelet/dockertools/manager.go#L1493) in the docker manager.

If the container PID is 0, we should check the ExitCode and other relevant fields and handle them appropriately.

[1] This is the suspected cause of several flakes:
http://kubekins.dls.corp.google.com/job/kubernetes-e2e-gke/7253/
http://kubekins.dls.corp.google.com/job/kubernetes-e2e-gce-container-vm/25/
http://kubekins.dls.corp.google.com/job/kubernetes-e2e-gce-container-vm/32/

/cc @kubernetes/sig-node @bprashanth @smarterclayton @zmerlynn

### LINKED ISSUE #25879: e2e flake: Pods should invoke init containers on a RestartAlways pod
https://console.cloud.google.com/storage/browser/kubernetes-jenkins/logs/kubernetes-e2e-gke/7294

```
• Failure [60.592 seconds]
[k8s.io] Pods
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:508
  should invoke init containers on a RestartAlways pod [It]
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/pods.go:794

  Expected
      <api.ConditionStatus>: False
  to equal
      <api.ConditionStatus>: True
```

@dchen1107 - is this a Node bug?
cc @lavalamp as next oncall to follow up on this.

### PR DESCRIPTION
**Pull request by Random-Liu**
Fix https://github.com/kubernetes/kubernetes/issues/25879
Fix https://github.com/kubernetes/kubernetes/issues/25844

This PR changed `convertToAPIContainerStatuses` to only generate waiting state when the init container really needs to be restarted.
Addresses https://github.com/kubernetes/kubernetes/issues/25844#issuecomment-220418068

Will send a better fix and add unit test later.
/cc @yujuhong @smarterclayton

### CODE REVIEW

File: pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -3671,6 +3671,14 @@ func (kl *Kubelet) convertToAPIContainerStatuses(pod *api.Pod, podStatus *kubeco
 
 	// Handle the containers failed to be started, which should be in Waiting state.
 	for _, container := range containers {
+		if hasInitContainers {
```
Comment (yujuhong) [inline pkg/kubelet/kubelet.go]: should use `isInitContainers`

File: pkg/kubelet/kubelet.go [hunk 2]
```diff
@@ -3671,6 +3671,14 @@ func (kl *Kubelet) convertToAPIContainerStatuses(pod *api.Pod, podStatus *kubeco
...
+			// If the init container is terminated with exit code 0, it won't be restarted.
+			// TODO(random-liu): Handle this in a cleaner way and add unit test for this.
```
Comment (yujuhong) [inline pkg/kubelet/kubelet.go]: Could you add a simple unit test for init containers in this PR? At least we can make sure the fix works :)
The test should be similar to the `TestGenerateAPIPodStatus...` tests.

File: pkg/kubelet/kubelet.go (hunk 1 above)
Comment (Random-Liu) [inline pkg/kubelet/kubelet.go]: The code above uses `hasInitContainers`, so I'm not sure which to use...

/cc @smarterclayton

Comment (smarterclayton) [inline pkg/kubelet/kubelet.go]: isInitContainer is for when we are iterating over the list of init containers (so rules that apply to init containers apply when that boolean is true).  This method is called for both init containers and regular containers.

Comment (Random-Liu) [inline pkg/kubelet/kubelet.go]: @smarterclayton Thanks, understood! :) And will add a unit test in this PR.

File: pkg/kubelet/kubelet.go (hunk 2 above)
Comment (Random-Liu) [inline pkg/kubelet/kubelet.go]: Will add a unit test in this PR.

**Comment by Random-Liu** [comment]
@yujuhong Unit test added.

**Comment by smarterclayton** [comment]
Lgtm too, thanks for tracking this down

**Comment by Random-Liu** [comment]
@k8s-bot test this issue: #25953

**Pull request merged**